modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php

```diff
@@ -220,53 +267,115 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
-      $form_state->setErrorByName('name', $this->t('Unrecognized username or password.'));
+        // Advanced validation Make sure the account does not exist yet. And
+        // that the username is unused/valid.
+        if ($this->entityTypeManager->getStorage('user')->loadByProperties(['mail' => $values['register']['mail']])) {
+          $form_state->setErrorByName('mail', $this->t('A user is already registered with this email.'));
+          return;
+        }
+        if ($this->entityTypeManager->getStorage('user')->loadByProperties(['name' => $values['register']['mail']])) {
+          $form_state->setErrorByName('mail', $this->t('A user is already registered with this username, please contact support to resolve this issue.'));
+          return;
+        }
+        // Make sure the email would be a valid username.
+        if (user_validate_name($values['register']['mail'])) {
+          $form_state->setErrorByName('mail', $this->t('The email you have used contains bad characters.'));
+          return;
+        }
+
+        // Create the new account.
+        $account = $this->entityTypeManager->getStorage('user')->create([]);
+        $account->setEmail($values['register']['mail']);
+        $account->setUsername($values['register']['mail']);
+        $account->setPassword($values['register']['pass']);
+        $account->enforceIsNew();
+        $account->activate();
+        $account->save();
+
+        // Login.
+        $form_state->set('logged_in_uid', $account->id());
+        drupal_set_message($this->t('Registration successful. You can now continue the checkout.'));
+        break;
+
+      default:
+        $form_state->setError($pane_form['returning_customer']['name'], $this->t('Invalid submission, please submit the form again.'));
+        break;
     }
-    $form_state->set('logged_in_uid', $uid);
   }
 
   /**
    * {@inheritdoc}
    */
   public function submitPaneForm(array &$pane_form, FormStateInterface $form_state, array &$complete_form) {
     $triggering_element = $form_state->getTriggeringElement();
-    if ($triggering_element['#op'] == 'login') {
-      $storage = $this->entityTypeManager->getStorage('user');
-      $account = $storage->load($form_state->get('logged_in_uid'));
-      user_login_finalize($account);
-      $this->order->setOwner($account);
-      $this->credentialsCheckFlood->clearAccount($this->clientIp, $account->getAccountName());
+
+    switch ($triggering_element['#op']) {
+      case 'login':
+      case 'register':
+        $storage = $this->entityTypeManager->getStorage('user');
+        /** @var \Drupal\user\UserInterface $account */
+        $account = $storage->load($form_state->get('logged_in_uid'));
+        user_login_finalize($account);
+        $this->order->setOwner($account);
+        $this->credentialsCheckFlood->clearAccount($this->clientIp, $account->getAccountName());
+        break;
+
+      case 'continue':
```

Comment:
Why have both the continue and the default cases, which do the same thing? Kill one